sklearn/metrics/_decision_threshold.py

```diff
@@ -0,0 +1,130 @@
+"""Metric per threshold curve to assess binary classification performance.
+
+Given threshold grid, one can undestand the behaviour of threshold-dependent
+metrics when changing the threshold. In imbalanced scenarios or
+cost-sensitive learning, a 0.5 threshold may not be optimal and tools like
+this can help you visualize how the performance changes.
+"""
+
+# Authors: The scikit-learn developers
+# SPDX-License-Identifier: BSD-3-Clause
+
+from numbers import Integral, Real
+
+from ..utils._param_validation import Interval, Options, validate_params
+
+
+@validate_params(
+    {
+        "scoring_function": [callable],
+        "y_true": ["array-like"],
+        "y_score": ["array-like"],
+        "thresholds": [
+            Interval(Integral, 2, None, closed="left"),
+            "array-like",
+        ],
+        "sign": Options(Real, {0, 1}),
+        "labels": ["array-like", None],
+        "pos_label": [Real, str, "boolean", None],
+    },
+    prefer_skip_nested_validation=True,
+)
+def decision_threshold_curve(
+    scoring_function,
+    y_true,
+    y_score,
+    # Should below 2 have a default value?
+    thresholds=20,
+    sign=1,
+    labels=None,
+    pos_label=None,
+    **kwargs,
+):
+    """Compute threshold-dependent metric of interest per threshold.
+
+    Note: this implementation is restricted to the binary classification task.
+
+    Read more in the :ref:`User Guide <metric_threshold_curve>`.
+
+    .. versionadded:: 1.8
+
+    Parameters
+    ----------
+    scoring_function : callable
+        The score function to use. It will be called as
+        `score_func(y_true, y_pred, **kwargs)`.
+        TODO: decided on `scoring_function` as term also used in forest estimators
+
+    y_true : array-like of shape (n_samples,)
+        Ground truth (correct) target labels.
+
+    y_score : array-like of shape (n_samples,)
+        Continuous response scores.
+
+    thresholds : int or array-like, default=20
+        Specifies number of decision thresholds to compute score for. If an integer,
+        it will be used to generate `thresholds` thresholds uniformly distributed
+        between the minimum and maximum of `y_score`. If an array-like, it will be
+        used as the thresholds.
+
+    sign : int, default=1
+        Either 1 or -1. Score is computed as `sign * score_func(estimator, X, y)`.
+        Thus, `sign` defines whether higher scores are better or worse.
+
+    labels: array-like, default=None
+        Class labels. If `None`, inferred from `y_true`.
+
+    pos_label : int, float, bool or str, default=None
```

Comment:
The other main question is whether we need to allow the user to set this, even when `scoring_function` does not take a `pos_label` parameter?

I *think* we should have this, because `pos_label` is passed to `_threshold_scores_to_class_labels`, and a user should be able to control this.